src/Sylius/Bundle/CoreBundle/Routing/Generator/ChannelProductUrlGenerator.php

```diff
@@ -0,0 +1,105 @@
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Routing\Generator;
+
+use Doctrine\Common\Collections\Collection;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ProductTranslationInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Locale\Model\LocaleInterface;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+final class ChannelProductUrlGenerator implements ChannelProductUrlGeneratorInterface
+{
+    public function __construct (
+        private LocaleContextInterface $localeContext,
+        private UrlGeneratorInterface $urlGenerator,
+        private bool $unsecuredUrls,
+    ) {
+    }
+
+    public function generate(ProductInterface $product, ChannelInterface $channel): ?string
+    {
+        /** @var Collection<array-key, ProductTranslationInterface> $productTranslations */
+        $productTranslations = $product->getTranslations();
+
+        $administratorLocaleCode = $this->localeContext->getLocaleCode();
+        $productTranslation = $this->findTranslationWithSlugForLocales($productTranslations, [$administratorLocaleCode]);
+
+        if (false !== $productTranslation) {
+            return $this->generateProductUrl($productTranslation, $channel);
+        }
+
+        /** @var string $channelDefaultLocaleCode */
+        $channelDefaultLocaleCode = $channel->getDefaultLocale()->getCode();
+        $productTranslation = $this->findTranslationWithSlugForLocales($productTranslations, [$channelDefaultLocaleCode]);
+
+        if (false !== $productTranslation) {
+            return $this->generateProductUrl($productTranslation, $channel);
+        }
+
+        $localesEnabledInChannel = $this->getLocalesCodesEnabledInChannel($channel);
+        $productTranslation = $this->findTranslationWithSlugForLocales($productTranslations, $localesEnabledInChannel);
+
+        if (false !== $productTranslation) {
+            return $this->generateProductUrl($productTranslation, $channel);
+        }
+
+        return null;
+    }
+
+    /**
+     * @param Collection<array-key, ProductTranslationInterface> $productTranslations
+     * @param array<string> $localeCodes
+     */
+    private function findTranslationWithSlugForLocales(Collection $productTranslations, array $localeCodes): ProductTranslationInterface|false
+    {
+        return $productTranslations->filter(function (ProductTranslationInterface $productTranslation) use ($localeCodes): bool {
+            return in_array($productTranslation->getLocale(), $localeCodes) &&
+                '' !== $productTranslation->getSlug() &&
+                null !== $productTranslation->getSlug()
+            ;
+        })->first();
+    }
```

Comment:
Instead of using a filter, you could do a foreach on the product translations and then do an early return without having to iterate through all of translations or return null.